File: src/py/ThriftClient/client_test.py
import sys
import time
from hypertable.thriftclient import *
from hyperthrift.gen.ttypes import *

try:
  client = ThriftClient("localhost", 38080)
  print "HQL examples"

  try:
    namespace = client.open_namespace("bad")
  except:
    print "Caught exception when tyring to open 'bad' namespace"

  namespace = client.open_namespace("test")
  res = client.hql_query(namespace, "show tables")
  print res
  res = client.hql_query(namespace, "select * from thrift_test")
  print res

  print "mutator examples";
  mutator = client.open_mutator(namespace, "thrift_test", 0, 0);
  client.set_cell(mutator, Cell(Key("py-k1", "col", None), "py-v1"))
  client.flush_mutator(mutator);
  
  print "shared mutator examples";
  mutate_spec = MutateSpec("test_py", 1000, 0);
  client.offer_cell(namespace, "thrift_test", mutate_spec, Cell(Key("py-put-k1", "col", None), "py-put-v1"))
  client.refresh_shared_mutator(namespace, "thrift_test", mutate_spec)
  client.offer_cell(namespace, "thrift_test", mutate_spec, Cell(Key("py-put-k2", "col", None), "py-put-v2"))
  time.sleep(2)

  print "scanner examples";
  scanner = client.open_scanner(namespace, "thrift_test",
                                ScanSpec(None, None, None, 1), True);

  while True:
    cells = client.next_cells(scanner)
    if (len(cells) == 0):
      break
    print cells
  
  print "asynchronous scanner examples\n";
  future = client.open_future(0);
  color_scanner = client.open_scanner_async(namespace, "FruitColor", future, ScanSpec(None, None, None, 1), True);
  location_scanner = client.open_scanner_async(namespace, "FruitLocation", future, ScanSpec(None, None, None, 1), True);
  energy_scanner = client.open_scanner_async(namespace, "FruitEnergy", future, ScanSpec(None, None, None, 1), True);

  
  expected_cells = 6;
  num_cells = 0;

  while True:
    result = client.get_future_result(future);
    print result;
    if (result.is_empty or result.is_error or not(result.is_scan) ):
      print "Unexpected result\n"
      exit(1); 
    for cell in result.cells:
      print cell;
      num_cells+=1;
    if(num_cells >= 6):
      client.cancel_future(future);
      break;

  print "regexp scanner example";
  scanner = client.open_scanner(namespace, "thrift_test",
      ScanSpec(None, None, None, 1, 0, None, None, ["col"], False,0, "k", "v[24]"), True);

  while True:
    cells = client.next_cells(scanner)
    if (len(cells) == 0):
      break
    print cells
  
  client.close_scanner_async(color_scanner); 
  client.close_scanner_async(location_scanner);
  client.close_scanner_async(energy_scanner);
  client.close_future(future);
  
  client.close_namespace(namespace)
except:
  print sys.exc_info()
  raise

File: src/perl/ThriftClient/client_test.pl
#!/usr/bin/env perl
use Hypertable::ThriftClient;
use Data::Dumper;

my $client = new Hypertable::ThriftClient("localhost", 38080);

print "HQL examples\n";
my $namespace = $client->open_namespace("test");
print Dumper($client->hql_exec($namespace,"show tables"));
print Dumper($client->hql_exec($namespace,"select * from thrift_test revs=1"));

print "mutator examples\n";
my $mutator = $client->open_mutator($namespace, "thrift_test");
my $key = new Hypertable::ThriftGen::Key({row => 'perl-k1',
                                          column_family => 'col'});
my $cell = new Hypertable::ThriftGen::Cell({key => $key,
                                            value => 'perl-v1'});
$client->set_cell($mutator, $cell);
$client->flush_mutator($mutator);

print "shared mutator examples\n";
my $mutate_spec = new Hypertable::ThriftGen::MutateSpec({appname => "test-perl",
                                                         flush_interval => 1000,
                                                         flags => 0});
$key = new Hypertable::ThriftGen::Key({row => 'perl-put-k1',
                                       column_family => 'col'});
$cell = new Hypertable::ThriftGen::Cell({key => $key,
                                         value => 'perl-put-v1'});
$client->offer_cell($namespace, "thrift_test", $mutate_spec, $cell);

$key = new Hypertable::ThriftGen::Key({row => 'perl-put-k2',
                                       column_family => 'col'});
$cell = new Hypertable::ThriftGen::Cell({key => $key,
                                         column_family => 'col',
                                         value => 'perl-put-v2'});
$client->refresh_shared_mutator($namespace, "thrift_test", $mutate_spec);
$client->offer_cell($namespace, "thrift_test", $mutate_spec, $cell);
sleep(2);

print "scanner examples\n";
my $scanner = $client->open_scanner($namespace, "thrift_test",
    new Hypertable::ThriftGen::ScanSpec({revs => 1}));

my $cells = $client->next_cells($scanner);

while (scalar @$cells) {
  print Dumper($cells);
  $cells = $client->next_cells($scanner);
}

print "asynchronous examples\n";
my $future = $client->open_future();
my $color_scanner = $client->open_scanner_async($namespace, "FruitColor", $future,
    new Hypertable::ThriftGen::ScanSpec({revs => 1}));
my $location_scanner = $client->open_scanner_async($namespace, "FruitLocation", $future,
    new Hypertable::ThriftGen::ScanSpec({revs => 1}));
my $energy_scanner = $client->open_scanner_async($namespace, "FruitEnergy", $future,
    new Hypertable::ThriftGen::ScanSpec({revs => 1}));

my $expected_cells = 6;
my $num_cells=0;

while (1) {
  my $result = $client->get_future_result($future);
  print Dumper($result);
  last if ($result->{is_empty}==1 || $result->{is_error}==1 || $result->{is_scan}!=1);
  my $cells = $result->{cells};
  foreach my $cell (@$cells){
    print Dumper($cell);
    $num_cells++;
  }
  if ($num_cells >= 6) {
    $client->cancel_future($future);
    last;
  }
}

$client->close_scanner_async($color_scanner);
$client->close_scanner_async($location_scanner);
$client->close_scanner_async($energy_scanner);;
$client->close_future($future);

die "Expected $expected_cells cells got $num_cells." if ($num_cells != $expected_cells);

print "regexp scanner example\n";
$scanner = $client->open_scanner($namespace, "thrift_test",
    new Hypertable::ThriftGen::ScanSpec({revs => 1, row_regexp=>"k", value_regexp=>"^v[24]",
    columns=>["col"]}));

my $cells = $client->next_cells($scanner);

while (scalar @$cells) {
  print Dumper($cells);
  $cells = $client->next_cells($scanner);
}

$client->close_namespace($namespace);
